File: 2023/java/D03P1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashSet;

public class d03p1 {
    private static final int ZERO = 48;
    private static final int NINE = 57;

    public static void main(String[] args) throws IOException {
        System.out.println("Day 3 - Part 1");

        BufferedReader reader = new BufferedReader(new FileReader("../_input/day03.txt"));
        int[][] lines = reader.lines().map(line -> line.chars().toArray()).toArray(int[][]::new);

        int total = process(lines);
        System.out.println("Total: " + total);
    }

    private static int process(final int[][] lines) {
        int total = 0;

        for(int row = 0; row < lines.length; row++) {
            int[] line = lines[row];
            for(int col = 0; col < line.length; col++) {
                int c = line[col];

                if(c >= ZERO && c <= NINE) {
                    HashSet<SymbolID> syms = new HashSet<>();
                    StringBuilder num = new StringBuilder();

                    while(col < line.length && c >= ZERO && c <= NINE) {
                        for(int rr = Math.max(row - 1, 0); rr < row + 2 && rr < lines.length; rr++) {
                            for(int cc = Math.max(col - 1, 0); cc < col + 2 && cc < line.length; cc++) {
                                int sc = lines[rr][cc];
                                if(!(sc >= ZERO && sc <= NINE) && sc != '.')
                                    syms.add(new SymbolID(rr, cc));
                            }
                        }
                        
                        num.append((char) c);
                        col++;
                        c = line[col < line.length ? col : 0];
                    }

                    int number = stringToInt(num.toString());
                    if(!syms.isEmpty()) total += number;
                }
            }
        }

        return total;
    }

    private static int stringToInt(String str) {
        int acc = 0;
        for(int i = 0; i < str.length(); i++) {
            int c = str.charAt(i);
            acc += (int) (Math.pow(10.0, str.length() - i - 1) * (c - ZERO));
        }
        return acc;
    }

    private record SymbolID(int row, int col) {
        @Override
        public int hashCode() {
            return row * col + row;
        }
    }
}

File: 2023/rust/src/bin/d03p1.rs
use std::fs;

fn main() {
    println!("Day 3 - Part 1");

    let str = fs::read_to_string("../_input/day03.txt").unwrap();
    let lines: Vec<&str> = str.lines().collect();

    let total = process(lines);
    println!("Total: {}", total);
}

fn process(lines: Vec<&str>) -> u32 {
    let mut l = Vec::new();
    for line in lines {
        l.push(line.chars().into_iter().collect::<Vec<char>>());
    }
    let lines = l;

    let mut total = 0;

    let mut i = 0;
    while i < lines.len() {
        let line = &lines[i];

        let mut num = String::new();
        let mut symbol = false;

        let mut li = 0;
        while li < line.len() {
            let c = line[li];

            if c.is_ascii_digit() {
                num.push(c);

                // Check
                if !symbol {
                    // Top Left
                    if i != 0
                        && li != 0
                        && !lines[i - 1][li - 1].is_ascii_digit()
                        && !lines[i - 1][li - 1].is_ascii_whitespace()
                        && !lines[i - 1][li - 1].eq(&'.')
                    {
                        symbol = true;
                    }
                    // Top Mid
                    if i != 0
                        && !lines[i - 1][li].is_ascii_digit()
                        && !lines[i - 1][li].is_ascii_whitespace()
                        && !lines[i - 1][li].eq(&'.')
                    {
                        symbol = true;
                    }
                    // Top Right
                    if i != 0
                        && li != line.len() - 1
                        && !lines[i - 1][li + 1].is_ascii_digit()
                        && !lines[i - 1][li + 1].is_ascii_whitespace()
                        && !lines[i - 1][li + 1].eq(&'.')
                    {
                        symbol = true;
                    }

                    // Mid Left
                    if li != 0
                        && !lines[i][li - 1].is_ascii_digit()
                        && !lines[i][li - 1].is_ascii_whitespace()
                        && !lines[i][li - 1].eq(&'.')
                    {
                        symbol = true;
                    }
                    // Mid Right
                    if li != line.len() - 1
                        && !lines[i][li + 1].is_ascii_digit()
                        && !lines[i][li + 1].is_ascii_whitespace()
                        && !lines[i][li + 1].eq(&'.')
                    {
                        symbol = true;
                    }

                    // Bottom Left
                    if i != lines.len() - 1
                        && li != 0
                        && !lines[i + 1][li - 1].is_ascii_digit()
                        && !lines[i + 1][li - 1].is_ascii_whitespace()
                        && !lines[i + 1][li - 1].eq(&'.')
                    {
                        symbol = true;
                    }
                    // Bottom Mid
                    if i != lines.len() - 1
                        && !lines[i + 1][li].is_ascii_digit()
                        && !lines[i + 1][li].is_ascii_whitespace()
                        && !lines[i + 1][li].eq(&'.')
                    {
                        symbol = true;
                    }
                    // Bottom Right
                    if i != lines.len() - 1
                        && li != line.len() - 1
                        && !lines[i + 1][li + 1].is_ascii_digit()
                        && !lines[i + 1][li + 1].is_ascii_whitespace()
                        && !lines[i + 1][li + 1].eq(&'.')
                    {
                        symbol = true;
                    }
                }
            }

            if !c.is_ascii_digit() || li + 1 >= line.len() {
                if symbol {
                    total += str_to_u32(&num);
                }

                num = String::new();
                symbol = false;
            }

            li += 1;
        }

        i += 1;
    }

    total
}

fn str_to_u32(str: &str) -> u32 {
    let mut acc = 0;
    for (i, c) in str.chars().enumerate() {
        acc += (10_u32.pow((str.len() - i - 1) as u32)) * c.to_digit(10).unwrap();
    }
    acc
}

#[cfg(test)]
mod test {
    use crate::process;

    #[test]
    fn test() {
        assert_eq!(
            process(
                r"467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598.."
                    .lines()
                    .collect()
            ),
            4361
        );
    }
}
